homeassistant/components/heos/coordinator.py

```diff
@@ -265,3 +283,64 @@ async def async_play_source(self, source: str, player: HeosPlayer) -> None:
             translation_key="unknown_source",
             translation_placeholders={"source": source},
         )
+
+    async def async_join_players(
+        self, player_id: int, member_entity_ids: list[str]
+    ) -> None:
+        """Join a player with a group of players."""
+        player_ids: list[int] = [player_id]
+        # Resolve entity_ids to player_ids
+        entity_registry = er.async_get(self.hass)
+        for entity_id in member_entity_ids:
+            entity = entity_registry.async_get(entity_id)
+            if entity is None:
+                raise ServiceValidationError(
```

Comment:
It should be the service handler that raises this exception. The controller method doesn't know it's part of a service handler. The controller method could raise some other exception that the service handler catches and raises as a `ServiceValidationError`.

Reply:
While I understand that reasoning, practically that doesn't make sense. This is a service action implementation and only ever used in that context. Defining custom exceptions (to differentiate the two scenarios currently raising) and then translate them at the single call site does not make sense. To resolve this, I could move this code into the `media_player` itself, or into a helper in `services.py` as it is not directly related to the update coordinator itself. Thoughts?

Reply:
I moved the group join/unjoin logic into the media_player for better architectural alignment and negates the need to define custom exceptions and map them.